Chart: Enhance Celery Worker Sets support for multi-queue configurations

glennhsh opened the pull request:
### Description
This PR enhances the Airflow Helm chart to support advanced Celery worker topologies, enabling more flexible resource allocation and precise autoscaling configurations.

### Why is this needed?

**1. Flexible Worker Topologies**
As Airflow adoption grows, platform teams often need to route tasks exclusively to specialized worker sets (e.g., GPU-optimized, Memory-optimized) without maintaining a generic "default" worker.
*   **Enhancement:** The new `workers.enableDefault` flag allows users to configure a deployment consisting *only* of specialized worker sets defined in `workers.sets`. This provides greater flexibility for teams to design their worker architecture exactly as needed.

**2. Multi-Queue Autoscaling Support**
Complex workflows often require a single worker set to handle tasks from multiple specific queues (e.g., `queue: "high-priority,vip"`).
*   **Enhancement:** This PR updates the KEDA `ScaledObject` generation to support comma-separated queue lists. By using the SQL `IN (...)` clause, we ensure that KEDA scales worker sets based on the precise aggregate workload of all their assigned queues.

**3. Granular Configuration Overrides**
Different worker sets may require different operational strategies within the same cluster.
*   **Enhancement:** This change improves the configuration merge logic, allowing individual worker sets to override global settings. For example, a user can now enable KEDA globally but explicitly disable it for a specific worker set that requires a static number of replicas.

### Changes
*   **New Feature:** Added `workers.enableDefault` (default: `true`) to `values.yaml`.
*   **Enhancement:** Updated `worker-kedaautoscaler.yaml` to use SQL `IN` clause for queue filtering, supporting multi-queue configurations (e.g., `queue: "a,b"` -> `AND queue IN ('a','b')`).
*   **Refactor:** Standardized template rendering to ensure consistent behavior between the default worker and `workers.sets`.

### Testing
*   Added test cases in `helm-tests/tests/helm_tests/other/test_keda.py` to verify:
    *   Correct SQL generation for single queues.
    *   Correct SQL generation for comma-separated queue lists using the `IN` clause.
    *   Proper handling of whitespace in queue configurations.
*   Verified that `workers.enableDefault` correctly controls the rendering of the default worker deployment.

closes: #56591 
closes: #34219

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

Commit:
Chart: Enhance Celery Worker Sets support for multi-queue configurations

This commit improves the configuration and scaling logic for Celery Worker Sets,
allowing for more robust multi-queue setups.

Key changes:
1.  **Strict KEDA Queue Filtering**: Updated KEDA SQL queries to always include
    `AND queue = '...'`. Previously, the default worker's KEDA query could
    incorrectly include tasks from other queues (e.g., those assigned to specific
    worker sets), leading to incorrect scaling behavior.

2.  **Explicit Default Worker Toggle**: Introduced `workers.enableDefault` (default: `true`).
    This allows users to easily disable the default worker if they wish to rely
    solely on custom worker sets for specific queues, improving configuration clarity.

3.  **Independent Resource Generation**: Refactored HPA and KEDA templates to
    generate resources for the default worker and worker sets independently.
    This resolves issues where the default worker's autoscaling resources were
    sometimes suppressed when sets were defined.

4.  **Test Updates**: Updated Helm tests to verify the new queue filtering logic
    and the independent generation of worker resources.

Commit:
Chart: Support multiple queues in KEDA autoscaling query

This change updates the KEDA SQL query generation in the Helm chart to
properly handle multiple queues defined in `workers.queue`.

By using `splitList` and iterating over the queues, the generated SQL
now uses an `IN` clause (e.g., `queue IN ('default', 'high-cpu')`)
instead of a simple equality check. This ensures that the Horizontal
Pod Autoscaler scales correctly when workers are listening to multiple
queues.

Commit:
Chart: Update schema defaults and fix lint errors

This change updates `values.schema.json` to reflect the recent changes
in `values.yaml` for KEDA multi-queue support.

Specific changes:
- Sync default values for `workers.args` and `workers.keda.query`.
- Fix `lint-chart-schema` failure by changing `additionalProperties: true`
  to `additionalProperties: {}` in `workers.sets`.

jscheffl reviewed (approved):
For me this looks good but would like another pair of eyes for review.

Miretpl commented on chart/templates/workers/worker-deployment.yaml:
```diff
@@ -20,10 +20,20 @@
 ################################
 ## Airflow Worker Deployment
 #################################
+{{- $workerSets := .Values.workers.sets | default list -}}
```
We will have two new sections under `workers` with the next release: `workers.kubernetes` and `workers.celery`. This change will only affect celery workers, so maybe it would be a good idea to move it there?

potiuk replied:
Yep.

glennhsh replied:
Thanks a lot for the review!
I completely agree with the direction. Since this change only impacts Celery workers, I’ve moved it under workers.celery as suggested.
Appreciate the clarification and guidance 🙏

Miretpl reviewed (commented):
Really nice change! Maybe it would be worth to add a newsfragment for this feature? 🤔

Commit:
Chart: Move Celery worker set configuration under workers.celery

Commit:
Chart: Add significant Helm chart newsfragment for multiple Celery worker sets

jscheffl commented on chart/newsfragments/58547.significant.rst:
```diff
@@ -0,0 +1,8 @@
+Users upgrading the chart should review their Helm values if they relied on the previous
+single worker behavior.
+
+See `Helm PR #58547 <https://github.com/apache/airflow/pull/58547>`_ for details.
```
Thanks for adding the newsfragment.

Instead of pointing to the PR description can you maybe add the details to the RST documentation? Let's assume in 125 years the project moves away from Github and Microsoft is acquired by Nvidia and then Nvidia decides to tear down Github... then documentation would be "lost". Would be better to make the documentation "proper" into the Helm chart docs.

glennhsh replied:
Thanks for the feedback and the suggestion!
That makes sense — I agree it’s better to have the details properly documented in the Helm chart docs rather than relying on the PR description. I’ll update the RST documentation accordingly. Thanks again for pointing this out.

jscheffl replied:
I was passing along this PR again, I think the small RST documentation is the only thing left over, then we can merge.

glennhsh replied:
Hi @jscheffl,
Thanks for your patience, and sorry for the delay.
I’ve updated the RST docs with the required details and incorporated the suggestions from the other contributors.
Feel free to review when you have a moment — happy to make further adjustments if needed.
Appreciate everyone’s input!

Miretpl commented on chart/values.yaml:
```diff
@@ -637,6 +637,12 @@ kerberos:
 # Airflow Worker Config
 workers:
+  # Enable the default worker defined by the root 'workers' configuration.
+  # If false, only workers defined in 'sets' will be created.
```
```suggestion
  # Enable the default workers defined by the root 'workers' configuration to be created.
  # If false, only dedicated workers defined in 'sets' will be created.
```
I feel like the description from one point is good and from the other something is missing for me, so just a small proposal

Miretpl commented on chart/values.yaml:
```diff
@@ -637,6 +637,12 @@ kerberos:
+  # If false, only workers defined in 'sets' will be created.
+  enableDefault: true
+
+  # Queue name for the worker
```
```suggestion
  # Queue name for the default workers
```

Miretpl reviewed (commented):
Despite the change in the newsfragment mentioned by Jens, everything looks good to me. Just small potential nits to some comments

glennhsh committed:
Update chart/values.yaml

Co-authored-by: Przemysław Mirowski <miretpl@gmail.com>

glennhsh committed:
Update chart/values.yaml

Co-authored-by: Przemysław Mirowski <miretpl@gmail.com>

glennhsh committed:
Enhance Helm chart for multiple Celery worker sets

This PR enhances the Airflow Helm chart to support advanced Celery worker topologies, enabling flexible resource allocation and precise autoscaling configurations. It introduces new features like workers.enableDefault, multi-queue autoscaling support, and granular configuration overrides for worker sets.

Miretpl commented on chart/newsfragments/58547.significant.rst:
```diff
@@ -0,0 +1,24 @@
+1. Flexible Worker Topologies
+As Airflow adoption grows, platform teams often need to route tasks exclusively to specialized worker sets (e.g., GPU-optimized, Memory-optimized) without maintaining a generic "default" worker.
+
+Enhancement: The new workers.enableDefault flag allows users to configure a deployment consisting only of specialized worker sets defined in workers.celery.sets. This provides greater flexibility for teams to design their worker architecture exactly as needed.
```
```suggestion
Enhancement: The new ``workers.enableDefault`` flag allows users to configure a deployment consisting only of specialized worker sets defined in ``workers.celery.sets``. This provides greater flexibility for teams to design their worker architecture exactly as needed.
```

Miretpl commented on chart/newsfragments/58547.significant.rst:
```diff
@@ -0,0 +1,24 @@
+
+Enhancement: This change improves the configuration merge logic, allowing individual worker sets to override global settings. For example, a user can now enable KEDA globally but explicitly disable it for a specific worker set that requires a static number of replicas.
+Changes
+New Feature: Added workers.enableDefault (default: true) to values.yaml.
```
```suggestion
New Feature: Added ``workers.enableDefault`` (default: true) to values.yaml.
```

Miretpl commented on chart/newsfragments/58547.significant.rst:
```diff
@@ -0,0 +1,24 @@
+Changes
+New Feature: Added workers.enableDefault (default: true) to values.yaml.
+Enhancement: Updated worker-kedaautoscaler.yaml to use SQL IN clause for queue filtering, supporting multi-queue configurations (e.g., queue: "a,b" -> AND queue IN ('a','b')).
+Refactor: Standardized template rendering to ensure consistent behavior between the default worker and workers.celery.sets.
```
```suggestion
Refactor: Standardized template rendering to ensure consistent behavior between the default ``workers`` and ``workers.celery.sets``.
```

Miretpl reviewed (approved):
Small nits in the newsfragment, despite that looks good.

glennhsh committed:
Update chart/newsfragments/58547.significant.rst

Co-authored-by: Przemysław Mirowski <miretpl@gmail.com>

glennhsh committed:
Update chart/newsfragments/58547.significant.rst

Co-authored-by: Przemysław Mirowski <miretpl@gmail.com>

glennhsh committed:
Update chart/newsfragments/58547.significant.rst

Co-authored-by: Przemysław Mirowski <miretpl@gmail.com>

Commit:
Fix worker set persistence.enabled value not being preserved

Helm's mustMerge and merge functions do not preserve boolean false
values because false is considered a "zero value" in Go templates.
This caused workers.celery.persistence.enabled=false to be
incorrectly overwritten by the deprecated workers.persistence.enabled
default value (true) during context merging.

The fix saves the persistence.enabled value before merge operations
and restores it afterward. Also corrects the reference path from
.Values.workers.celery.persistence.enabled to
.Values.workers.persistence.enabled since the worker configuration
is placed under .Values.workers in the new context.

Miretpl commented on helm-tests/tests/helm_tests/airflow_core/test_worker.py:
```diff
@@ -1519,7 +1519,7 @@ def test_worker_template_storage_class_name(self, workers_values):
             ({"celery": {"replicas": None}}, 1),
             ({"replicas": 2, "celery": {"replicas": 3}}, 3),
-            ({"replicas": 2, "celery": {"replicas": None}}, 2),
+            ({"replicas": 2, "celery": {"replicas": 2}}, 2),
```
With moving from `workers.replicas` to `workers.celery.replicas`, we want to make sure that when users unset the `workers.celery.replicas` field, the behaviour will be as in previous releases (because we changed how chart behaves in replicas handling). Why was this change made?

I see now locally that this was changed as the default behaviour was changed. `workers.replicas`, if `workers.celery.replicas` is unset, does't change the value of replicas. I think it should be fixed in the template logic, not in the test case itself

Miretpl replied:
I think I've managed to fix it. @glennhsh, could you do a review of the change https://github.com/apache/airflow/pull/60420? For making sure that `workers.celery.sets` are working correctly, I've only used the current unit tests
